datafusion/optimizer/src/eliminate_filter.rs

```diff
@@ -78,17 +56,47 @@ impl OptimizerRule for EliminateFilter {
     fn apply_order(&self) -> Option<ApplyOrder> {
         Some(ApplyOrder::TopDown)
     }
+
+    fn supports_rewrite(&self) -> bool {
+        true
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>> {
+        match plan {
+            LogicalPlan::Filter(Filter {
+                predicate: Expr::Literal(ScalarValue::Boolean(v)),
+                input,
+                ..
+            }) => match v {
+                Some(true) => Ok(Transformed::yes(
+                    self.rewrite(unwrap_arc(input), _config)?.data,
```

Comment:
I almost wonder if this recursion is even necessary given that it already does `ApplyOrder::TopDown` the optimizer itself will call `Self::rewrite` on the inputs

In other words, it might be possible to do

```suggestion
                    unwrap_arc(input)
```